cloudinit/distros/__init__.py

```diff
@@ -239,29 +239,39 @@ def install_packages(self, pkglist: PackageList):
 
         # First install packages using package manager(s)
         # supported by the distro
-        uninstalled = []
+        uninstalled = set()
         for manager in self.package_managers:
-            to_try = (
-                packages_by_manager.get(manager.__class__, set())
-                | generic_packages
+
+            manager_packages = packages_by_manager.get(
+                manager.__class__, set()
             )
+
+            to_try = manager_packages | generic_packages
+            uninstalled.difference_update(to_try)
+            if not manager.available():
+                LOG.debug("Package manager '%s' not available", manager.name)
+                uninstalled.update(to_try)
+                continue
             if not to_try:
                 continue
-            uninstalled = manager.install_packages(to_try)
-            failed = {
-                pkg for pkg in uninstalled if pkg not in generic_packages
+            failed = manager.install_packages(to_try)
+            uninstalled.update(failed)
```

Comment:
Shouldn't this be the following?
```suggestion
            uninstalled = to_try.difference(failed)
```

Reply:
I changed the wording from `uninstalled` to `total_failed` to hopefully be a little more clear.

> Shouldn't this be the following?

No, doing this won't keep track of failures from previous package managers.

Say we have
```yaml
packages:
- pkg1
- apt:
  - pkg2
- snap:
  - pkg3
``` 

If `pkg2` failed, it will stay failed regardless of happens during snap install, so overwriting `total_failed` each time in the loop won't work. On the other hand, if `pkg1` fails during apt install, we will still attempt it during snap install, so what happened during apt install should get replaced by what happened during snap. The logic is a bit convoluted, so I'm open to any ideas for simplification. 

This is the reason for the new `test_failed` test in `distros/test_init.py`